tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+  k_star: int
+  n: int
+  dtype: tf.DType
+
+  def F(self, z1, z2, tau):
+    """Calculates F(z1, z2) (eq. (11) from the referenced paper)"""
+    x = tf.math.sqrt(tf.math.pow(self.mean_reversion, 2)
+                     - 2 * z1 * tf.math.pow(self.volvol, 2))
+
+    term_1 = tf.math.cosh(0.5 * tau * x)
+    term_2 = ((self.mean_reversion - z2 * tf.math.pow(self.volvol, 2))
+               * tf.math.sinh(0.5 * tau * x)) / x
+
+    return term_1 + term_2
+
+  def F_tilde(self, z1, z2, tau):
+    """Calculates F_tilde(z1, z2) (eq. (11) from the referenced paper)"""
+    x = tf.math.sqrt(tf.math.pow(self.mean_reversion, 2)
+                     - 2 * z1 * tf.math.pow(self.volvol, 2))
+
+    term_1 = 0.5 * x * tf.math.sinh(0.5 * tau * x)
+    term_2 = 0.5 * ((self.mean_reversion - z2 * tf.math.pow(self.volvol, 2))
+                     * tf.math.cosh(0.5 * tau * x))
+
+    return term_1 + term_2
+
+  def a(self, s, w):
+    """Calculates a(s, w) (eq. (17) from the referenced paper)"""
+    raw_drift = self.discount_rates - self.dividend_rates
+    corr_drift = self.rho * self.mean_reversion * self.theta / self.volvol
+    mod_drift = raw_drift - corr_drift
+
+    summed_samplings = tf.expand_dims(
+      tf.reduce_sum(self.sampling_times[self.k_star:], axis=0),
+      -1
+    )
+
+    n_c = tf.constant(self.n, dtype=self.dtype) 
+    k_star_c = tf.constant(self.k_star, dtype=self.dtype)
+
+    term_1 = ((s * (n_c - k_star_c) / n_c + w)
+               * (tf.math.log(self.spots)
+                  - self.rho * self.variances / self.volvol
+                  - self.start_times * mod_drift))
+    term_2 = (s * (summed_samplings) / n_c + w * self.expiries) * mod_drift
+
+    return term_1 + term_2
+
+  def z(self, s, w):
+    """Calculates z_k(s, w) (eq. (14) from the referenced paper)"""
+    k_s_w_tensor = tf.expand_dims(self.n - self.k_tensor + 1, 2) * s + self.n * w
+    term_1 = ((2 * self.rho * self.mean_reversion - self.volvol)
+               * (k_s_w_tensor) / (self.volvol * self.n))
+    term_2 = ((1 - tf.math.pow(self.rho, 2))
+               * tf.cast(tf.math.pow(k_s_w_tensor, 2), dtype=self.dtype)
+               / tf.cast(tf.math.pow(self.n, 2), dtype=self.dtype))
+
+    return 0.5 * (term_1 + term_2)
+
+  def omega_k(self, s, w):
+    """Calculates \omega_k(s, w) (eq. (15) from the referenced paper)"""
+    return tf.concat([tf.expand_dims(self.start_times, 1) * s,
+                      tf.expand_dims(tf.ones((self.sampling_times.shape[1], self.n - self.k_star), dtype=self.dtype) * self.rho / (self.volvol * self.n), 2) * s,
+                      tf.expand_dims(self.rho / self.volvol, 1) * w], 1)
+
+  def omega_tilde_k(self, s, w, z_k_tensor):
+    """Calculates \tilde{\omega}_k(s, w) (eq. (19) from the referenced paper)"""
+    omega_tilde = self.omega_k(s, w)
+    mr_v2 = self.mean_reversion / tf.math.pow(self.volvol, 2)
+
+    rows = []
+    last_row = None
+
+    for q in reversed(range(0, omega_tilde.shape[1])):
```

Comment:
seems like this can be vecotrized. could you please try using `tf.vecorized_map`  over q here?

Reply:
Ahhh this is a tricky issue. \tilde{\omega} is calculated recursively (it's where the Recursive in the paper's title comes from actually), I've reproduced it here:

![image](https://user-images.githubusercontent.com/10614716/207342737-753e2d1c-d7ef-4075-81a1-39e68e752fd9.png)

You can see it's very serial - each row must be calculated and then plugged into the next row's calculation in a very non-linear fashion. I tried lots of tricks to untangle it but haven't been able to so far, in the end this loop formulation was the clearest thing I could settle on.

I'd love to have any thoughts on ways we might manage this better. For example, definitely the final row would need to come after all of the preceding rows have been calculated, but we *could* start any part of the calculation that only depends on the first x rows as soon as those have been calculated - if there are any tricks you can think of here I'd love to see them.

Reply:
ok, this makes sense. It is better to manually add tf.while_loop inside the library as otherwise, when [Autograph](https://www.tensorflow.org/api_docs/python/tf/autograph) is disabled, this may unroll to an unnecessary large graph.

So something similar to [Euler sampling](https://github.com/google/tf-quant-finance/blob/7a8ed52958fc21a168d8fd419a339008f521d846/tf_quant_finance/models/euler_sampling.py#L405):

```
num_iterations = utils.get_shape(omega_tilde)[1]

rows = tf.TensorArray(dtype=dtype,
                            size=num_iterations,
                            element_shape='shape of omega?',
                            clear_after_read=Fals)

def cond(it, omega_tilde, rows):
  return  it < num_iterations

def body(it, omega_tilde, rows):
   ...
  row = ...
  ...
  rows = rows.write(it, row)
  return it + 1, omega_tilde, rows

_, _, rows = tf.while_loop()

rows = rows.stack()
```

Please let me know if you need help filling in the details

Reply:
Done! It works nicely and the tests are passing. Thanks for this feedback, it's much nicer than the original for loop solution.